## Week2 Assignment   
Welcome to week2 assignment notebook! This is a place that I practice coding for week2 materials 😀


### A quick Geopandas teaser
I will start by importing geopandas to the notebook

In [1]:
import geopandas as gpd

Next, I will import a dataset. I uploaded data to the notebook.

In [5]:
bike = gpd.read_file('data/metro-trips-2020-q4.csv')

I want to check data type.

In [6]:
type(bike)

geopandas.geodataframe.GeoDataFrame

I want to check how the data looks by checking the first five rows.

In [7]:
bike.head()

trip_id duration      start_time        end_time start_station  \
0  148179433       14  10/1/2020 0:06  10/1/2020 0:20          3042   
1  148179933       18  10/1/2020 0:25  10/1/2020 0:43          4404   
2  148182639      141  10/1/2020 0:29  10/1/2020 2:50          4482   
3  148182739      141  10/1/2020 0:29  10/1/2020 2:50          4482   
4  148180233       15  10/1/2020 0:42  10/1/2020 0:57          3074   

   start_lat    start_lon end_station    end_lat      end_lon bike_id  \
0  34.049301  -118.238808        3074   34.04417  -118.261169    6378   
1   34.04813  -118.271027        4444  34.061619  -118.305573   12440   
2  34.094372  -118.331009        4482  34.094372  -118.331009   19803   
3  34.094372  -118.331009        4482  34.094372  -118.331009   18915   
4   34.04417  -118.261169        3042  34.049301  -118.238808   20062   

  plan_duration trip_route_category passholder_type bike_type geometry  
0            30             One Way    Monthly Pass  standard     None  
1            30             One Way    Monthly Pass  standard     None  
2             1          Round Trip         Walk-up  electric     None  
3             1          Round Trip         Walk-up  electric     None  
4            30             One Way    Monthly Pass  standard     None

and also the last five rows..

In [9]:
bike.tail()

trip_id duration        start_time          end_time start_station  \
37872  151711987      126  12/31/2020 23:00     1/1/2021 1:06          4210   
37873  151709683       15  12/31/2020 23:06  12/31/2020 23:21          3052   
37874  151710184        6  12/31/2020 23:42  12/31/2020 23:48          4516   
37875  151710183        6  12/31/2020 23:42  12/31/2020 23:48          4516   
37876  151710983       39  12/31/2020 23:46     1/1/2021 0:25          4401   

       start_lat    start_lon end_station    end_lat      end_lon bike_id  \
37872  33.984341   -118.47155        4210  33.984341   -118.47155    6551   
37873  34.051102  -118.264557        4518  34.057968  -118.299751    6221   
37874  34.045849  -118.254402        3052  34.051102  -118.264557   12305   
37875  34.045849  -118.254402        3052  34.051102  -118.264557    6476   
37876  34.075611  -118.291862        4401  34.075611  -118.291862    6038   

      plan_duration trip_route_category passholder_type bike_type geometry  
37872             1          Round Trip    One Day Pass  standard     None  
37873            30             One Way    Monthly Pass  standard     None  
37874             1             One Way         Walk-up  standard     None  
37875             1             One Way         Walk-up  standard     None  
37876             1          Round Trip    One Day Pass  standard     None

### Pandas data types

In [10]:
bike.dtypes

trip_id                  object
duration                 object
start_time               object
end_time                 object
start_station            object
start_lat                object
start_lon                object
end_station              object
end_lat                  object
end_lon                  object
bike_id                  object
plan_duration            object
trip_route_category      object
passholder_type          object
bike_type                object
geometry               geometry
dtype: object

Moving to next step to better see the data types.

In [12]:
bike.info()

<class 'geopandas.geodataframe.GeoDataFrame'>
RangeIndex: 37877 entries, 0 to 37876
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype   
---  ------               --------------  -----   
 0   trip_id              37877 non-null  object  
 1   duration             37877 non-null  object  
 2   start_time           37877 non-null  object  
 3   end_time             37877 non-null  object  
 4   start_station        37877 non-null  object  
 5   start_lat            37877 non-null  object  
 6   start_lon            37877 non-null  object  
 7   end_station          37877 non-null  object  
 8   end_lat              37877 non-null  object  
 9   end_lon              37877 non-null  object  
 10  bike_id              37877 non-null  object  
 11  plan_duration        37877 non-null  object  
 12  trip_route_category  37877 non-null  object  
 13  passholder_type      37877 non-null  object  
 14  bike_type            37877 non-null  object  
 15  geometry   

### Data exploration

In [13]:
bike.shape

(37877, 16)

I want to see the header.

In [14]:
bike.columns

Index(['trip_id', 'duration', 'start_time', 'end_time', 'start_station',
       'start_lat', 'start_lon', 'end_station', 'end_lat', 'end_lon',
       'bike_id', 'plan_duration', 'trip_route_category', 'passholder_type',
       'bike_type', 'geometry'],
      dtype='object')

In [15]:
bike.columns.to_list()

['trip_id',
 'duration',
 'start_time',
 'end_time',
 'start_station',
 'start_lat',
 'start_lon',
 'end_station',
 'end_lat',
 'end_lon',
 'bike_id',
 'plan_duration',
 'trip_route_category',
 'passholder_type',
 'bike_type',
 'geometry']

I want to know how many trips were generated and ended in each station.

In [19]:
bike['start_station'].value_counts()

3005    1589
4210    1128
3035    1015
3006     874
4491     816
        ... 
4573      11
4500      11
4569      10
4559       9
4432       1
Name: start_station, Length: 200, dtype: int64

In [20]:
bike['end_station'].value_counts()

3005    1575
4210    1090
3000    1029
3035     964
4215     928
        ... 
4569      11
4559      10
4432       9
4416       8
4500       6
Name: end_station, Length: 200, dtype: int64


**We can see that station 3005 and 4210 is top ranked in both the originatiion and destination for bike trips**

**(3005 is '7th & Flower' station and 4210 is 'Ocean Front Walk & North Venice' station)**

I want to save this as a variable. From now on I will conduct same code for both the starting station(origination) and ending station(destination).

In [21]:
S_station_count = bike['start_station'].value_counts()
E_station_count = bike['end_station'].value_counts()

In [22]:
S_station_count = S_station_count.reset_index()
S_station_count

index  start_station
0    3005           1589
1    4210           1128
2    3035           1015
3    3006            874
4    4491            816
..    ...            ...
195  4573             11
196  4500             11
197  4569             10
198  4559              9
199  4432              1

[200 rows x 2 columns]

In [23]:
E_station_count = E_station_count.reset_index()
E_station_count

index  end_station
0    3005         1575
1    4210         1090
2    3000         1029
3    3035          964
4    4215          928
..    ...          ...
195  4569           11
196  4559           10
197  4432            9
198  4416            8
199  4500            6

[200 rows x 2 columns]

### Rename columns

In [24]:
S_station_count.columns.to_list()

['index', 'start_station']

In [26]:
S_station_count.columns = ['station', 'count']

In [27]:
S_station_count

station  count
0      3005   1589
1      4210   1128
2      3035   1015
3      3006    874
4      4491    816
..      ...    ...
195    4573     11
196    4500     11
197    4569     10
198    4559      9
199    4432      1

[200 rows x 2 columns]

In [28]:
E_station_count.columns.to_list()
E_station_count.columns = ['station', 'count']
E_station_count

station  count
0      3005   1575
1      4210   1090
2      3000   1029
3      3035    964
4      4215    928
..      ...    ...
195    4569     11
196    4559     10
197    4432      9
198    4416      8
199    4500      6

[200 rows x 2 columns]

In [ ]:
query, plot 만들기 